src/Compilers/CSharp/Portable/CodeGen/EmitOperators.cs

```diff
@@ -59,36 +59,119 @@ private void EmitBinaryOperatorExpression(BoundBinaryOperator expression, bool u
 
             if (operatorKind.EmitsAsCheckedInstruction())
             {
-                EmitBinaryCheckedOperatorExpression(expression, used);
+                EmitBinaryOperator(expression);
+            }
+            else
+            {
+                // if operator does not have side-effects itself and is not short-circuiting
+                // we can simply emit side-effects from the first operand and then from the second one
+                if (!used && !operatorKind.IsLogical() && !OperatorHasSideEffects(operatorKind))
+                {
+                    EmitExpression(expression.Left, false);
+                    EmitExpression(expression.Right, false);
+                    return;
+                }
+
+                if (IsConditional(operatorKind))
+                {
+                    EmitBinaryCondOperator(expression, true);
+                }
+                else
+                {
+                    EmitBinaryOperator(expression);
+                }
+            }
+
+            EmitPopIfUnused(used);
+        }
+
+        private void EmitBinaryOperator(BoundBinaryOperator expression)
+        {
+            BoundExpression child = expression.Left;
+
+            if (child.Kind != BoundKind.BinaryOperator || child.ConstantValue != null)
+            {
+                EmitBinaryOperatorSimple(expression);
                 return;
             }
 
-            // if operator does not have side-effects itself and is not short-circuiting
-            // we can simply emit side-effects from the first operand and then from the second one
-            if (!used && !operatorKind.IsLogical() && !OperatorHasSideEffects(operatorKind))
+            BoundBinaryOperator binary = (BoundBinaryOperator)child;
+            var operatorKind = binary.OperatorKind;
+
+            if (!operatorKind.EmitsAsCheckedInstruction() && IsConditional(operatorKind))
```

Comment:
I agree with Aleksey.  Unless perf says this is a bad idea I'd prefer to have a single approach to generating these.